freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/KerberosMgmtV1Service.java

```diff
@@ -149,12 +160,45 @@ private String serviceAdd(String serviceName, String hostname, String realm, Str
                 }
                 service = ipaClient.showService(canonicalPrincipal);
             }
-            allowServiceKeytabRetrieval(service.getKrbprincipalname(), adminUser, ipaClient);
+            allowServiceKeytabRetrieval(service.getKrbprincipalname(), freeIpaClientFactory.getAdminUser(), ipaClient);
+            addRoleAndPrivileges(service, request.getRoleRequest(), ipaClient);
         } catch (FreeIpaClientException e) {
             LOGGER.error(SERVICE_PRINCIPAL_CREATION_FAILED + " " + e.getLocalizedMessage(), e);
             throw new KeytabCreationException(SERVICE_PRINCIPAL_CREATION_FAILED);
         }
-        return service.getKrbprincipalname();
+        return service;
+    }
+
+    private void addRoleAndPrivileges(com.sequenceiq.freeipa.client.model.Service service, RoleRequest roleRequest, FreeIpaClient ipaClient)
+            throws FreeIpaClientException {
+        if (roleRequest != null && StringUtils.isNotBlank(roleRequest.getRoleName())) {
+            Set<Role> allRole = ipaClient.findAllRole();
+            Optional<Role> optionalRole = allRole.stream().filter(role -> role.getCn().equals(roleRequest.getRoleName())).findFirst();
+            Role role = optionalRole.isPresent() ? optionalRole.get() : ipaClient.addRole(roleRequest.getRoleName());
+            addPrivilegesToRole(roleRequest.getPrivileges(), ipaClient, role);
+            role = ipaClient.showRole(role.getCn());
+            boolean roleSetForService = service.getMemberOfRole().stream().anyMatch(member -> member.contains(roleRequest.getRoleName()));
+            if (!roleSetForService) {
+                ipaClient.addRoleMember(role.getCn(), null, null, null, null, Set.of(service.getKrbprincipalname()));
+            }
+        }
+    }
+
+    private void addPrivilegesToRole(Set<String> privileges, FreeIpaClient ipaClient, Role role) throws FreeIpaClientException {
+        if (privileges != null) {
+            Set<String> privilegesToAdd = privileges.stream().filter(privilegeName -> {
```

Comment:
Can you please explain the intent of this block? 
I think member in the privilege would be roles, right. What is the logic looking for "privilegeName" in members of privilege with name "privilegeName".

Reply:
good catch! That should be the roleName